src/nfa/nfa-concatenation.cc

```diff
@@ -26,6 +26,43 @@ Nfa concatenate(const Nfa& lhs, const Nfa& rhs, bool use_epsilon,
     return Algorithms::concatenate_eps(lhs, rhs, EPSILON, use_epsilon, lhs_result_states_map, rhs_result_states_map);
 }
 
+Nfa& Nfa::concatenate(const Nfa& aut) {
+    size_t n = this->size();
+    auto upd_fnc = [&](State st) {
+        return st + n;
+    };
+
+    this->delta.append(aut.delta.transform(upd_fnc));
+
+    // set accepting states
+    Util::SparseSet<State> new_fin{};
+    new_fin.reserve(n+aut.size());
+    for(const State& aut_fin : aut.final) {
+        new_fin.insert(upd_fnc(aut_fin));
+    }
+
+    // connect both parts
+    for(const State& ini : aut.initial) {
+        const Post& ini_post = this->delta[upd_fnc(ini)];
+        // is ini state also final?
+        bool is_final = aut.final[ini];
```

Comment:
isn't is too much? just wandering

Reply:
If you check something in a for loop multiple times, it is always worth extracting the test outside the for loop, in my opinion. Even if the optimizer will most probably do the same thing as we do. We do the same with the condition check in the for loop itself. If we check for, for example, `i < delta.size()`, we rewrite it as `const size_t delta_size = delta.size(); for (size_t i{ 0 }; i < delta_size; ++i)`.